File: hash_table/0599_minimum_index_sum_of_two_lists/0599_minimum_index_sum_of_two_lists.go
package main

import (
	"fmt"
)

func findRestaurant(list1 []string, list2 []string) []string {
	if len(list1) > len(list2) {
		list1, list2 = list2, list1
	}

	m2 := make(map[string]int, len(list2))
	for i := range list2 {
		m2[list2[i]] = i
	}

	min := 2000

	res := make([]string, 0, 1000)
	for i, r := range list1 {
		if j, ok := m2[r]; ok {
			if min == i+j {
				res = append(res, r)
			}
			if min > i+j {
				res = append(res[len(res):], r)
				min = i + j
			}
		}
	}
	return res
}

func main() {
	list1 := []string{"Shogun", "Tapioca Express", "Burger King", "KFC"}
	list2 := []string{"Piatti", "The Grill at Torrey Pines", "Hungry Hunter Steakhouse", "Shogun"}
	fmt.Println(findRestaurant(list1, list2))
}

File: hash_table/0599_minimum_index_sum_of_two_lists/0599_minimum_index_sum_of_two_lists.py
class Solution(object):
    def findRestaurant(self, list1, list2):
        """
        :type list1: List[str]
        :type list2: List[str]
        :rtype: List[str]
        """
        lookup = {}
        for i, s in enumerate(list1):
            lookup[s] = i

        result = []
        min_sum = float("inf")
        for j, s in enumerate(list2):
            if j > min_sum:
                break
            if s in lookup:
                if j + lookup[s] < min_sum:
                    result = [s]
                    min_sum = j + lookup[s]
                elif j + lookup[s] == min_sum:
                    result.append(s)
        return result

list1 = ["Shogun","Tapioca Express","Burger King","KFC"]
list2 = ["Piatti","The Grill at Torrey Pines","Hungry Hunter Steakhouse","Shogun"]
print(Solution().findRestaurant(list1, list2))